### ISSUE TITLE
Use login shell for executing commands inside of containers

### ISSUE THREAD

**Issue by felipecrs** [microsoft/vscode-remote-release#3593]
After starting the devcontainer, `vscode` executes commands inside of the container by some sort of:

```sh-session
$ docker exec devcontainer sh -c "command to execute"
```

An example of a command is to start the `code-server`.

This feature request is to change the above `docker exec` call to something like:

```sh-session
$ docker exec devcontainer sh -l -c "command to execute"
```

Notice the `-l` option. It makes the shell to `source` the `/etc/profile` before executing the command, which enables us to customize the container environment by following Linux standards and best practices for doing such a thing.

This enables us to define global environment variables using scripts and not only through the `Dockerfile`'s `ENV` statement, or the `devcontainer.json`'s `containerEnv`.

This would mainly benefit the scripts from `script-library` on `vscode-dev-containers` repository, which are used for doing common customizations on the containers. As explained [here](https://github.com/microsoft/vscode-dev-containers/issues/512#issuecomment-680426014) (bullet number 2).

**Comment by chrmarti** [comment]
Does `"userEnvProbe"` in the devcontainer.json not solve this?

**Comment by felipecrs** [comment]
I'll test it soon and let you know. Thanks.

**Comment by felipecrs** [comment]
@chrmarti Yes, this is what I was looking for, but with a caveat.

Without `"userEnvProbe": "loginInteractiveShell"` it does not work:

![Code_xQdtxS8Zr1](https://user-images.githubusercontent.com/29582865/91916654-09166900-ec94-11ea-9e0a-c7b13f4db972.png)

But with `"userEnvProbe": "loginInteractiveShell"` it works as expected:

![Code_5GAzbW2qIt](https://user-images.githubusercontent.com/29582865/91916710-2ba88200-ec94-11ea-8efe-42f4cb0b0b95.png)

However, I was actually looking for an option `loginShell` and not `loginInteractiveShell`, which translates to `sh -l` and `sh -l -i` respectively.

And, I'm proposing to make `loginShell` option as the default value instead of `none`. The reason is:

> This would mainly benefit the scripts from `script-library` on `vscode-dev-containers` repository, which are used for doing common customizations on the containers. As explained [here](https://github.com/microsoft/vscode-dev-containers/issues/512#issuecomment-680426014) (bullet number 2).

**Comment by chrmarti** [comment]
`"userEnvProbe"` doesn't affect the building of the Docker image. Couldn't you change your Dockerfile to use `RUN bash -l -c "..."` if using a login shell allows you to shorten it?

Taking this as the feature request to add `"loginShell"` for `"userEnvProbe"`.

**Comment by PavelSosin-320** [comment]
Why it is better than entry-point script? Container is not multitenant. Repetitive login can create sibling processes that will be never killed by init during the termination and become ghost OS processes.  This is very dangerous.

**Comment by felipecrs** [comment]
> "userEnvProbe" doesn't affect the building of the Docker image. Couldn't you change your Dockerfile to use RUN bash -l -c "..." if using a login shell allows you to shorten it?

This is not about that, this is about setting up container wide environment not by using `ENV`. By the way, if someone wants to do what you said, it's better to use `SHELL [ "bash", "-l", "-c" ]`.

**Comment by felipecrs** [comment]
> Why it is better than entry-point script?

I don't understand how this is related to the entry point script. Remember that entry point script is not run on `docker exec`. [I actually want the entry point to be executed all the time, but it's a totally different history.](https://github.com/microsoft/vscode-remote-release/issues/3614).

> Container is not multitenant. Repetitive login can create sibling processes that will be never killed by init during the termination and become ghost OS processes. This is very dangerous.

Do you have any example? I use login shell in my Linux environments since ever, and I don't see anything harm in `/etc/profile`. Even if something could be found (such as starting a process), this block would probably check if there is a command already started. [Such as the `ssh-agent`](https://stackoverflow.com/a/18915067/12156188).

Also, the login environment would be only applied to the command which we are executing, it would not affect other running shells for example.

That said, do you have any reference for what you said?

**Comment by Chuxel** [comment]
@chrmarti I'm actually a +1 on doing this by default.  There's actually quite a bit of code in vscode-dev-containers that is working around this particular scenario.  The big problem is scripts that must be sourced to get the path to work correctly.  SDKMAN, nvm, rvm all are like this.  I'm adding things the containers install to the PATH when possible, but this only covers a limited set. In the SDKMAN case there are a number of tools it can install. Worse yet, rvm throws up warnings if things are in the path and the targets are not easily predictable (e.g. the gem path).  To your point, we could probably add SHELL to work around it, but I could see someone trying to create their own dev container hitting this and being confused.

**Comment by chrmarti** [comment]
The issue with `/etc/profile` is that in some images (e.g., Alpine and Debian) it overwrites `PATH` (investigation https://github.com/microsoft/vscode-remote-release/issues/3224#issuecomment-650034161), so additions to PATH via `ENV` are lost (bug https://github.com/microsoft/vscode-remote-release/issues/544). Due to that we changed to run with no startup scripts sourced.

One option is to probe the environment with an interactive non-login shell (`"userEnvProbe": "interactiveShell"`), but that adds to the startup time (numbers on my machine https://github.com/microsoft/vscode-remote-release/issues/3224#issuecomment-656032893) so we decided to not do that by default.

**Comment by chrmarti** [comment]
@Chuxel Are there any workarounds that cannot be replaced with the recently added `"userEnvProbe"`?

**Comment by Chuxel** [comment]
@chrmarti I don't believe so. I need to do a more complete test. This last release there was enough change already that I opted to keep things as they were.

**Comment by chrmarti** [comment]
@Chuxel If it turns out that you have to use `"userEnvProbe": "interactiveShell"` on almost all definitions we should reconsider the default.

I expect `"loginShell"` `"loginInteractiveShell"` to be used less often due to some `/etc/profile` overwriting `PATH`. If those would be better choices, we could think about somehow merging the two `PATH`s. That is not trivial because the order of PATH entries matters and we might need another property for the user to configure (or turn off) that - so we need good reasons to (also) take that approach.

**Comment by felipecrs** [comment]
The problem of using "interactiveShell" is (as the name says) that the shell is interactive, which means that at any point, bash or zsh can prompt the user for something. Not ideal in this case.

I would still go with the "loginShell" as default, and properly document how the users of Alpine containers (or others with this known issue of /etc/profile and PATH) can work around it by setting to "false".

I doubt alpine users would be the majority.

**Comment by Chuxel** [comment]
@felipecrs Typically CLIs are smart enough to determine when they should prompt even in an interactive shell due to piping.  As @chrmarti mentions, `.profile` can do all sorts of crazy things and typically also fires `.bashrc` under the hood, so of the two, `interactive` generally gives the expected results.

**Comment by felipecrs** [comment]
@Chuxel 'typically' is not a rule, right? The rule is: interactive terminals may ask the user for input. But it will probably work, as you suggest.

And the `/etc/bash.bashrc` is only sourced by `/etc/profile` if running in an interactive shell (that's why it checks for `$PS1`).

By the way, you're talking about `.profile` and `.bashrc`, but I suppose you are referring to `/etc/profile` and `/etc/bash.bashrc` and not `~/.profile` and `~/.bashrc`, right?

In the end, I [again] suggest using the standard way to customize the system's environment on Linux: adding as many `.sh` files you want in the `/etc/profile.d/` rather than editing `/etc/bash.bashrc` or even `/etc/profile` directly.

**Comment by felipecrs** [comment]
I also don't understand what "all sorts of crazy things" mean:

```sh-session
$ docker run --rm alpine sh -c 'cat /etc/profile; cat /etc/profile.d/*.sh'
cat: can't open '/etc/profile.d/*.sh': No such file or directory
export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
export PAGER=less
export PS1='\h:\w\$ '
umask 022

for script in /etc/profile.d/*.sh ; do
        if [ -r $script ] ; then
                . $script
        fi
done
```

```sh-session
$ docker run --rm debian sh -c 'cat /etc/profile; cat /etc/profile.d/*.sh'
# /etc/profile: system-wide .profile file for the Bourne shell (sh(1))
# and Bourne compatible shells (bash(1), ksh(1), ash(1), ...).

if [ "`id -u`" -eq 0 ]; then
  PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
else
  PATH="/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games"
fi
export PATH

if [ "${PS1-}" ]; then
  if [ "${BASH-}" ] && [ "$BASH" != "/bin/sh" ]; then
    # The file bash.bashrc already sets the default PS1.
    # PS1='\h:\w\$ '
    if [ -f /etc/bash.bashrc ]; then
      . /etc/bash.bashrc
    fi
  else
    if [ "`id -u`" -eq 0 ]; then
      PS1='# '
    else
      PS1='$ '
    fi
  fi
fi

if [ -d /etc/profile.d ]; then
  for i in /etc/profile.d/*.sh; do
    if [ -r $i ]; then
      . $i
    fi
  done
  unset i
fi
cat: '/etc/profile.d/*.sh': No such file or directory
```

```sh-session
$ docker run --rm ubuntu sh -c 'cat /etc/profile; cat /etc/profile.d/*.sh'
# /etc/profile: system-wide .profile file for the Bourne shell (sh(1))
# and Bourne compatible shells (bash(1), ksh(1), ash(1), ...).

if [ "${PS1-}" ]; then
  if [ "${BASH-}" ] && [ "$BASH" != "/bin/sh" ]; then
    # The file bash.bashrc already sets the default PS1.
    # PS1='\h:\w\$ '
    if [ -f /etc/bash.bashrc ]; then
      . /etc/bash.bashrc
    fi
  else
    if [ "`id -u`" -eq 0 ]; then
      PS1='# '
    else
      PS1='$ '
    fi
  fi
fi

if [ -d /etc/profile.d ]; then
  for i in /etc/profile.d/*.sh; do
    if [ -r $i ]; then
      . $i
    fi
  done
  unset i
fi
# Make sure the locale variables are set to valid values.
eval $(/usr/bin/locale-check C.UTF-8)
```

It seems that the only problem is the`PATH`.

**Comment by Chuxel** [comment]
@felipecrs Yep! That's why the property exists - so you can adapt to your situation.

**Comment by felipecrs** [comment]
@Chuxel Yeah, this is true. But the point of this issue is to come up with a better default approach, not only for my situation.

But I think the concerns about relying on the `/etc/profile` (`loginShell`) are right. Fixing up `PATH` variable does not seem right to me, too intrusive, which makes this option not feasible.

And as I said, I don't think that relying on `interactiveShell` is also a good idea, because of the reasons I pointed. Also, this would make `bashrc` to be sourced twice if the user opens a new `bash` under VSCode.

So I think I have a better proposal:

A new option in `devcontainer.json`:

```jsonc
{
  "loadDevContainerEnv": true // defaults to true
}
```

This option makes the Remote - Containers extension always execute commands in the container using the following syntax:

```sh-session
$ docker exec ubuntu sh -c "test -f /etc/devcontainer-env && . /etc/devcontainer-env; echo 'hello world'"
```

Unless the user sets it to `false`. This will not interfere in `userEnvProbe` behavior.

This would enable us to have a clean and unique environment configuration file that would work across all the users (`root` and non-root), and all the shells, without the caveats of `/etc/profile` or `bash.bashrc`.

@Chuxel @chrmarti could you please review this new proposal?

**Comment by chrmarti** [comment]
Note that the 'userEnvProbe' only starts a corresponding shell and uses `printenv` to extract the environment variables, it does not use that shell for anything else. So `bashrc` is sourced twice, but not in the same shell or a subshell. That should not cause the trouble you anticipate above. VS Code is using the same approach locally (except there it is always an interactive login shell and you cannot disable it).

We would probably use the same approach (`printenv`) to extract the environment variables if we added support for `/etc/devcontainer`, so we only source it once and are not forced to always use a shell.

**Comment by smhc** [comment]
Looks like a duplicate of https://github.com/microsoft/vscode-remote-release/issues/3585

**Comment by felipecrs** [comment]
> Looks like a duplicate of https://github.com/microsoft/vscode-remote-release/issues/3585

True, pretty much the same. I'll update the PR title and description to match https://github.com/microsoft/vscode-remote-release/issues/3593#issuecomment-690826874, so they'll be different.

**Comment by felipecrs** [comment]
I rewrote the PR.

**Comment by felipecrs** [comment]
Another thing, `BASH_ENV` environment variable can be used to specify bash rc files that should run if not in interactive or login mode. So, one could set in `Dockerfile` like:

```Dockerfile
ENV BASH_ENV="/etc/devcontainer"
SHELL ["/bin/bash", "-c"]
RUN echo "/etc/devcontainer will be sourced"
```

As it's an `ENV`, it will remain in the container and it will be used even when `"envProbe": "none"` (which is the default). It will not affect the terminals opened inside of VS Code, as `BASH_ENV` does not get executed when in interactive mode. This means that it's still needed to do something like:

```Dockerfile
RUN printf '%s\n' '' 'test -f /etc/devcontainer && . /etc/devcontainer' | sudo tee /etc/bash.bashrc /etc/zsh/zshrc
```

**Comment by QAston** [comment]
One problem with userEnvProbe is that it's not available for containers you attach to using attach to existing container option https://github.com/microsoft/vscode-remote-release/issues/6226

**Comment by chrmarti** [comment]
@QAston You can open the config file associated with the 'attach' container using `F1` > `Remote-Containers: Open Configuration File` and add `"userEnvProbe"` to that. The default is `"loginInteractiveShell"`.

**Comment by QAston** [comment]
@chrmarti thanks, I didn't notice it because it's not documented in the reference: https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_attached-container-configuration-reference

**Comment by chrmarti** [comment]
@QAston Thanks, opened a PR for the documentation. (https://github.com/microsoft/vscode-docs/pull/5119)
